src/NodaTime/TimeInterval.cs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2009 The Noda Time Authors. All rights reserved.
+// Use of this source code is governed by the Apache License 2.0,
+// as found in the LICENSE.txt file.
+
+using JetBrains.Annotations;
+using NodaTime.Text;
+using System;
+using NodaTime.Utility;
+
+namespace NodaTime
+{
+    /// <summary>
+    /// An interval between two local time values (start and end).
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The TimeInterval includes the start LocalTime and excludes the end LocalTime. However, an TimeInterval
```

Comment:
The part about a missing start time feels slightly odd to me, because it's exactly equivalent to having a start time of 00:00:00. (This isn't true for Interval.) So we can make sure there's *always* a start time... we can't do the same for the end time, as the end is exclusive and we don't support 24:00 as a `LocalTime` value.

With start being nullable, we can have two time intervals (one starting at 00:00:00 and one having no start) which are functionally equivalent, but not equal to each other. I'd quite like some more input about this.

@malcolmr @lundmikkel @mj1856 any thoughts?

Reply:
Actually I cannot imagine any case when the TimeInterval may not have a start or end. I want to suggest this logic: `if(start == null) start = LocalTime.MinValue`, `if(end == null) end = LocalTime.MaxValue`. Or maybe we should add `LocalTime.AfterMaxValue` to include the last nanosecond (because comparison at the moment is exclusively)? Or just make `end` comparison inclusively?
Please consider it.

Reply:
Not having a start is equivalent to midnight. That's easy. But it's entirely reasonable to create an interval of "Any time from 7pm onwards" but we can't represent that without something like LocalTime.AfterMaxValue, and I don't want to create that *and expose it*. (We've got that for Instant because it makes many things a lot simpler, but we avoid ever exposing it.)

Reply:
Then I suggest to make `start` not nullable, but leave `end` nullable.
So, Contains became:
```csharp
            if (EndsInNextDay)
            {
                if (check < Start) return check < End;
                return true;
            }
            return check >= Start && (!end.HasValue || check < End);
```
And Duration became:
```csharp
                var endNanoseconds = end.HasValue ? End.NanosecondOfDay : NodaConstants.NanosecondsPerDay;
                var nanoseconds = endNanoseconds - Start.NanosecondOfDay;
                if (EndsInNextDay) nanoseconds += NodaConstants.NanosecondsPerDay;
                return Duration.FromNanoseconds(nanoseconds);
```

Reply:
Something like that, yes. But I think it's worth focusing on the API design before worrying too much about the implementation changes.

Reply:
Only one problem with this approach - TimeInterval of 22:00/00:00 and 22:00/null will not be equal, but internally they are. So, if we will use LocalTime.MinValue instead of null for end - it'll not work properly too, because of 00:00/00:00 TimeInterval will be interval with 0 Duration, not 24 hour as we except using null as end. What do you think about it? Should we modify Equals method too to make equal intervals like 22:00/00:00, 22:00/null?

Reply:
I hadn't thought that of course an end of midnight is "everything before midnight".

If it weren't for the fact that we couldn't represent "all day" as well as "no time at all" it would be perfect just to have Start and End as LocalTime.

Humbug. I think we really need to think carefully about how important it is to represent all the appropriate situations - and being able to construct them directly. (If we didn't mind "All day" or "Empty" being impossible to construct directly, I'm sure we could work out some cunning plan.)

(This is were designing an API for the library is much more time-consuming for designing a type that's useful within a single application, of course.)

I'll see whether java.time has anything like this, and what their approach was, if so.

Reply:
I don't see any LocalTime's intervals in java. Anyways, we can create 00:00/LocalTime.MaxValue and in fact it'll "all day". So, I suggest to make `end` non-nullable and just use MinValue for null-end. In comments we can note that null/null interval will became 00:00/00:00 interval.
I hope another developers will come up some additional use-cases to help us take a decision what logic to choose. 
Should I commit code changes now or lets solve this problem first?

Reply:
No, 00:00-LocalTime.MaxValue wouldn't mean "all day" - it would mean "all day except for the last nanosecond". Fundamentally we need to decide whether X-X (for any value of X) means "all day" or "empty". If we definitely need to support both of those use cases, we won't be able to *just* store two `LocalTime` values.

I'm genuinely not sure which of those use cases is more useful, or whether both are required.

If we're moving away from nullable ends though (which I'd generally like) then `LocalTime?` shouldn't appear anywhere in the API - we wouldn't need to note anything about a "null/null interval" because there'd be no such concept.

Reply:
So you suggest just remove all nullable types from TimeInterval? Including a nullable constructor? `public TimeInterval(LocalTime? start, LocalTime? end)`

Reply:
Yes, with the downside that we then can't represent both "empty" and "all day" - we need to decide which of those is the meaning when start and end are the same.

Reply:
Hello, I would like to propose another variant of the constructor that might be more useful:
```C#
public TimeInterval(LocalTime start, Duration duration)
{
    this.start = start;
    this.end = start.Plus(Period.FromNanoseconds((long)duration.TotalNanoseconds));
    IsEndInNextDay = start > end || (start == end && duration > Duration.Zero);
}
```
So when duration is zero then this interval is empty.

Reply:
That's certainly an interesting option. I think at that point I'd probably store this as a Start and Duration though, and expose that via properties (as well as an End still, of course). We could probably store the duration just as a number of nanoseconds, for efficiency.